File: python/src/topology/boundary_paths.cpp
/*
 * Copyright (c) 2025 Žiga Sajovic, XLAB
 * Licensed for noncommercial use under the PolyForm Noncommercial
 * License 1.0.0. Commercial licensing available via ziga.sajovic@xlab.si.
 * https://github.com/xlabmedical/trueform
 */

#include "trueform/python/topology/boundary_paths.hpp"
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>

namespace tf::py {

auto register_topology_boundary_paths(nanobind::module_ &m) -> void {
  using namespace nanobind;

  // ==========================================================================
  // BOUNDARY_PATHS
  // Index types: int32, int64
  // Ngon: 3, 4 (triangles, quads only)
  // Total: 4 bindings
  // ==========================================================================

  // int32, ngon=3 (triangles)
  m.def(
      "boundary_paths_int_3",
      [](ndarray<numpy, int, shape<-1, 3>> cells,
         const offset_blocked_array_wrapper<int, int> &fm) {
        return boundary_paths<int, 3>(cells, fm);
      },
      arg("cells"), arg("face_membership"));

  // int32, ngon=4 (quads)
  m.def(
      "boundary_paths_int_4",
      [](ndarray<numpy, int, shape<-1, 4>> cells,
         const offset_blocked_array_wrapper<int, int> &fm) {
        return boundary_paths<int, 4>(cells, fm);
      },
      arg("cells"), arg("face_membership"));

  // int64, ngon=3 (triangles)
  m.def(
      "boundary_paths_int64_3",
      [](ndarray<numpy, int64_t, shape<-1, 3>> cells,
         const offset_blocked_array_wrapper<int64_t, int64_t> &fm) {
        return boundary_paths<int64_t, 3>(cells, fm);
      },
      arg("cells"), arg("face_membership"));

  // int64, ngon=4 (quads)
  m.def(
      "boundary_paths_int64_4",
      [](ndarray<numpy, int64_t, shape<-1, 4>> cells,
         const offset_blocked_array_wrapper<int64_t, int64_t> &fm) {
        return boundary_paths<int64_t, 4>(cells, fm);
      },
      arg("cells"), arg("face_membership"));
}

} // namespace tf::py

File: python/trueform/_topology/boundary_paths.py
"""
boundary_paths() function implementation

Copyright (c) 2025 Ziga Sajovic, XLAB
Licensed for noncommercial use under the PolyForm Noncommercial License 1.0.0.
Commercial licensing available via ziga.sajovic@xlab.si.
https://github.com/xlabmedical/trueform
"""

import numpy as np
from .. import _trueform
from .._core import OffsetBlockedArray
from .._spatial import Mesh


def boundary_paths(mesh: Mesh) -> OffsetBlockedArray:
    """
    Extract boundary paths from a mesh.

    Connects boundary edges into paths (loops). Each path is a connected
    sequence of boundary edges forming a closed loop around a hole or
    the mesh exterior.

    Parameters
    ----------
    mesh : Mesh
        The mesh to extract boundary paths from.

    Returns
    -------
    OffsetBlockedArray
        Boundary paths where each block is a connected loop of vertex indices.
        Vertex indices reference the original mesh points.
        Returns empty OffsetBlockedArray if mesh has no boundaries.

    Examples
    --------
    >>> import trueform as tf
    >>> import numpy as np
    >>>
    >>> # Create a mesh with a hole
    >>> faces = np.array([
    ...     [0, 1, 2],
    ...     [0, 2, 3]
    ... ], dtype=np.int32)
    >>> points = np.array([
    ...     [0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0]
    ... ], dtype=np.float32)
    >>> mesh = tf.Mesh(faces, points)
    >>>
    >>> paths = tf.boundary_paths(mesh)
    >>> print(f"Found {len(paths)} boundary loops")
    >>>
    >>> # Iterate over boundary loops
    >>> for i, path in enumerate(paths):
    ...     print(f"Loop {i}: {len(path)} vertices")
    ...     # path contains original mesh vertex indices
    """

    # Validate input
    if not isinstance(mesh, Mesh):
        raise TypeError(
            f"mesh must be Mesh, got {type(mesh).__name__}"
        )

    # Validate ngon - only triangles and quads
    ngon = mesh.ngon
    if ngon not in (3, 4):
        raise ValueError(
            f"mesh must have triangular or quad faces, got {ngon} vertices per face. "
            f"Boundary detection is only defined for triangles and quads."
        )

    # Get faces and face_membership from mesh
    faces = mesh.faces
    fm = mesh._wrapper.face_membership_array()

    # Build suffix and dispatch
    dtype_str = 'int' if faces.dtype == np.int32 else 'int64'
    suffix = f"{dtype_str}_{ngon}"

    func_name = f"boundary_paths_{suffix}"
    cpp_func = getattr(_trueform.topology, func_name)

    # Call C++ function - returns offset_blocked_array_wrapper
    wrapper = cpp_func(faces, fm)

    return OffsetBlockedArray(wrapper.offsets_array(), wrapper.data_array())
